README.md
# NAME

Data::BoxFormat - Blah blah blah

# SYNOPSIS

    use Data::BoxFormat;

# DESCRIPTION

Data::BoxFormat is

# AUTHOR

Joseph Brenner <doom@kzsu.stanford.edu>

# COPYRIGHT

Copyright 2016- Joseph Brenner

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for Data-BoxFormat

{{$NEXT}}
        - Initial release


scripts/skullplot.el
;;; skullplot.el --- graphical plots of sql select results

;; Copyright 2016 Joseph Brenner
;;
;; Author: doom@kzsu.stanford.edu
;; Version: $Id: skullplot.el,v 0.0 2016/07/16 23:33:00 doom Exp $
;; Keywords:
;; X-URL: not distributed yet

;; License: the same as GNU Emacs, currently GPL v.3  (see LEGAL below).

;; Description:
;; A typical database shell (e.g. psql, mysql) displays the
;; results of queries in a tabular text format I call a "databox"
;; or "dbox" (file extension: *.dbox).

;; The skullplot.el package exists to assist in generating graphical
;; plots of SQL query results in this "dbox" form form (by the way:
;; database purists dislike pronouncing SQL as 'sequel', an obvious
;; alternative is 'skull').  It works with a perl script skullplot.pl
;; and a perl module Data::BoxFormat, and the R statistical language package
;; ggplot2.

;; (( TODO: every piece should check that all the other pieces are installed. ))

;;  The main user function here is:
;;   skullplot-of-this-dbox

;; Installation:

;; Put this file into your load-path and the following into your ~/.emacs:
;;   (require 'skullplot)

;; TODO key assignment for skullplot-of-this-dbox in various shells and db interfaces...

;;; Code:

(provide 'skullplot)
(eval-when-compile
  (require 'cl))

;;---------
;; early function definitions

(defun skullplot-fixdir (dir &optional root)
  "Fixes the DIR.
Conditions directory paths for portability and robustness.
Some examples:
 '~/tmp'             => '/home/doom/tmp/'
 '~/tmp/../bin/test' => '/home/bin/test/'
Note: converts relative paths to absolute, using the current
default-directory setting, unless specified otherwise with the
ROOT option, this has the side-effect of converting the empty
string into the default-directory or the ROOT setting."
  (let ((return
         (substitute-in-file-name
          (convert-standard-filename
           (file-name-as-directory
            (expand-file-name dir root))))))
    return))

;;---------
;; settings and globals

(defvar progname "skullplot"
  "Name of this program (used in warnings and so on).")

(defvar skullplot-clean-up-buffers t
  "Clean-up the temporary buffers used to pass dbox data.
Defaults to t, set to nil for debugging.")
(setq skullplot-clean-up-buffers nil)

(defvar skullplot-dbox-location ""
  "Location to place the *.dbox files")
(setq skullplot-dbox-location "~/.skullplot/dbox")
(skullplot-fixdir skullplot-dbox-location)

(defvar skullplot-hr-pattern ""
  "Matches 'horizontal rule' lines used in dbox format.
This matches lines like:
 ----+------------+-----------+--------
and also the psql unicode lines:
 ────┼────────────┼───────────┼────────
" )
(setq skullplot-hr-pattern "^[-─+┼][-─+┼][-─+┼]+$")

;; TODO
;; Something like pg format might have a *single* veritcal bar
;; if you've just done a select on two fields (and actually,
;; *none* is possible, if you select just one field... but
;; you're unlikely to want to plot that, correct?)
(defvar skullplot-vbar-pattern ""
  "Match lines that have any vertical bars on them." )
(setq skullplot-vbar-pattern "^.*?[|│]")

(defvar skullplot-vbar-or-cross-pattern ""
  "Heuristic match for any dbox lines, either data, header, or hrs." )
(setq skullplot-vbar-or-cross-pattern "^.*?[|│+┼]")

;; unused
(defvar skullplot-blank-line-pattern ""
  "Match a blank line." )
(setq skullplot-blank-line-pat "^[ \t]*$")


;;---------
;; dbox plotters
;; the main user command(s)

;; wrapper to use as entry point (?)  TODO
(defun skullplot-basic-group-by ( indie-count )
  "Takes a numeric argument to distinguish dependent columns from independent.
See section 'TODO'.  The argument defaults to 1."
  (interactive "p")
  ;; (message "indie-count: %d" indie-count )
  (skullplot-of-this-dbox indie-count)
 )

(defun skullplot-of-this-dbox ( indie-count )
  "Pop a window with a graphical display of data in the current
or immediately previous dbox."
  (interactive "p")
  (let* (
         ;; emacs uses exec-path list, which has the PATH added to it, maybe,
         ;; skip the use of progdir then...  pushing it onto end of PATH in .bashrc for now (bleh)
;;         (progdir "/home/doom/End/Cave/SkullPlot/Wall/Data-BoxFormat/scripts") ;; DEBUG
         (progfile "skullplot.pl")
;;         (slash "/") ;; TODO windows portability?
;;         (plotter      (concat progdir slash progfile) )
         (plotter    progfile)
;;          (tmpdir       "/home/doom/tmp")  ;; TODO (a) more specific? (b) create if not there
;;          (tmpfile      (concat tmpdir slash "skullplot.dbox") ) ;; TODO generate unique name...?
         (tmpfile (skullplot-create-unique-tempfile))
         (cmd (concat plotter " " tmpfile) )
           buffy
        )
    (setq table (skullplot-get-dbox))
    (find-file tmpfile)
    ;; clear the buffer
    (mark-whole-buffer)
    (delete-region (mark) (point)) ;; TODO maybe add safety features to preserve any existing content
    (insert table)
    (save-buffer)
    (setq buffy (buffer-name))
    (delete-window)
    (if skullplot-clean-up-buffers
         (kill-buffer buffy))

    ;; run the cmd asynchronusly in another process (no shell):
    (message "Running: %s" cmd);; DEBUG
    (start-process "skullplot" "*skullplot*" plotter tmpfile)
    ))

(defun skullplot-of-region (beg end)
  "Pop a window with a graphical display of data in dbox in region.
You might use this if the skullplot-of-this-dbox heuristics fail,
and you want to specify what to plot manually."
  (interactive "r") ;; r -- Region: point and mark as 2 numeric args, smallest first.  Does no I/O.

  (let* (
         ;; TODO just let the PATH sort it out?  but does system-command use PATH?
         ;; emacs uses exec-path list, which appears to have PATH added.
         ;; One approach: push progdir into exec-path.
         (progdir "/home/doom/End/Cave/SkullPlot/Wall/Data-BoxFormat/scripts")
         (progfile "skullplot.pl")
         (slash "/") ;; TODO windows portability?
         (plotter      (concat progdir slash progfile) )
         (tmpdir       "/home/doom/tmp")  ;; TODO (a) more specific? (b) create if not there
         (tmpfile      (concat tmpdir slash "skullplot.dbox") )
         (cmd (concat plotter " " tmpfile) )
        )
    (setq table (buffer-substring-no-properties beg end))
    (find-file tmpfile)
    ;; clear the buffer
    (mark-whole-buffer)
    (delete-region (mark) (point))
    ;;   TODO extra credit: add safety features, e.g. preserve any existing content

    (insert table)

    (save-buffer)
    ;;(write-file nil)
    ;; TODO close the buffer?  Or just the window?

    ;; run the cmd in another process (no shell):
    (message "Running: %s" cmd)

    ;; wasn't working, not sure why:
    ;; (start-process "skullplot" "*skullplot*" cmd)

    ;; works, but it blocks:
    ;; (call-process plotter tmpfile '("*skullplot*" t) t)

    ;; asynchronus
    (start-process "skullplot" "*skullplot*" plotter tmpfile)
))

;;---------
;; move to pattern

;;--------
;; hr rule

(defun skullplot-move-up-to-hr ()
  "Move upwards to first match of 'skullplot-hr-pattern' or start of buffer."
  (interactive)
  ;; Q: why not re-search-backward?  Eaiser to negate the pattern match, right?
  (move-beginning-of-line 1)
  (catch 'FLY
    (while (not (looking-at skullplot-hr-pattern ) )
      (forward-line -1) ;; move up to previous line
      (if (= (point) (point-min))
          (throw 'FLY nil))
      )))

(defun skullplot-move-down-to-hr ()
  "Move downwards to first match of 'skullplot-hr-pattern' or end of buffer.
Returns nil if not found."
  (interactive)
  ;; Q: why not re-search-forward?  A: eaiser to negate the pattern match, right?
  (move-beginning-of-line 1)
  (catch 'FLY
    (while (not (looking-at skullplot-hr-pattern ) )
      (forward-line 1)
      (move-end-of-line 1)
      (if (= (point) (point-max))
          (throw 'FLY nil))
      (move-beginning-of-line 1)
     )))

;;--------
;; vb sep lines or hrs

(defun skullplot-move-down-to-end-of-dbox ()
  "Move downwards to last line with vertical bars."
  (interactive)
  (move-beginning-of-line 1)
  (catch 'FLY
    (while ( looking-at skullplot-vbar-or-cross-pattern )
      (forward-line 1)
      (move-end-of-line 1)
      (if (= (point) (point-max))
          (throw 'FLY nil))
      (move-beginning-of-line 1)
     ))
  (forward-line -1)
  (move-end-of-line 1) )


(defun skullplot-move-up-to-start-of-dbox ()
  "Move upwards to first line with vertical bars."
  (interactive)
  (move-beginning-of-line 1)
  (catch 'FLY
    (while ( looking-at skullplot-vbar-or-cross-pattern )
      (forward-line -1) ;; move up to previous line
      (if (= (point) (point-min))
          (throw 'FLY nil))
      ))
  (forward-line 1))

;;--------
;; scraping dboxes

(defun skullplot-find-dbox ()
  "Look upwards for the previous dbox (if any)."
  (interactive)
  (let* ( (savepoint (point))
          dbox-start dbox-end  ;; start of upper line; end of lower line
          )
    (skullplot-move-up-to-hr)
    (skullplot-move-up-to-start-of-dbox)
    (setq dbox-start (point))
    (skullplot-move-down-to-end-of-dbox)
    (move-end-of-line 1)
    (setq dbox-end (point))
    ;; (message "%d %d" dbox-start dbox-end)
    (goto-char savepoint)
    (list dbox-start dbox-end)
    ))

(defun skullplot-get-dbox ()
  "Return previous dbox as string.."
  (interactive)
  (let* (( pair (skullplot-find-dbox) )
         ( beg (nth 0 pair) )
         ( end (nth 1 pair) )
         )
    (buffer-substring-no-properties beg end)))

;;--------
;; utilites

(defun skullplot-snag-line ()
  "Returns the current line as a string."
  (let ( start end line )
    (move-beginning-of-line 1)
    (setq start (point))
    (move-end-of-line 1)
    (setq end (point))
    (setq line (buffer-substring start end))
    line))

(defun skullplot-create-unique-tempfile ( &optional seed )
  "Create a uniquely named dbox file, using the SEED.
Location defaults to skullplot-dbox-location."
  (unless seed (setq seed ""))
  (let ( slashy temploc prefix templfile )
    (setq slashy (convert-standard-filename "/"))
    (setq temploc skullplot-dbox-location)
    (unless (file-exists-p temploc)
      (make-directory temploc t))
    (setq prefix
          (concat temploc slashy seed))
    (setq tempfile
          (make-temp-file prefix nil ".dbox"))
    )
  tempfile)



;; LEGAL

;; This code is licensed in the same way as GNU Emacs, which is currently
;; under the GPL v.3 (or later).

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

;;; skullplot.el ends here


scripts/skullplot.pl
#!/usr/bin/perl
# skullplot.pl                   doom@kzsu.stanford.edu
#                                27 Jul 2016

=head1 NAME

skullplot.pl - plot data from a manual db select

=head1 SYNOPSIS

  # default to first column on x-axis, all following columns on y-axis
  skullplot.pl input_data.dbox

  # specifying that explicitly
  skullplot.pl --dependents='x_axis_field' --independents='y_axis_field1,y_axis_field2' input_data.dbox

  # additional "dependents" fields determine color/shape of points
  skullplot.pl --dependents='x_axis_field1,category_field1' --independents='y_axis_field1,y_axis_field2' input_data.dbox

  # compact way of specifying similar case: first two columns independent, remaining columns dependent
  skullplot.pl --indie_count=2 input_data.dbox

  # don't use /tmp as working area
  skullplot.pl --working_loc='/var/scratch'  input_data.dbox

  # turn on debugging
  skullplot.pl -d  input_data.dbox

=head1 DESCRIPTION

B<skullplot.pl> is a script which use's the R ggplot2 library to
plot data input in the form of the output from a SELECT as
performed manually in a db shell, e.g.:

  +------------+---------------+-------------+
  | date       | type          | amount      |
  +------------+---------------+-------------+
  | 2010-09-01 | factory       |   146035.00 |
  | 2010-10-01 | factory       |   208816.00 |
  | 2011-01-01 | factory       |   191239.00 |
  | 2010-09-01 | marketing     |   467087.00 |
  | 2010-10-01 | marketing     |   409430.00 |
  +------------+---------------+-------------+

I call this "box format data" (file extension: *.dbox).

This script takes the name of the *.dbox file containing the data as an argument.
It has a number of options that control how it uses the data


The second argument is a comma-separated list of names of dependent variables
(the x-axis).
The third argument is a comma-separated list of the independent variables to
plot (the y-axis).

The default for dependent variables: the first column.
The default of independent variables: all of the following columns

The supported input data formats are as in the L<Data::BoxFormat> module.
At present, this is mysql and postgresql (including the unicode form).

=cut

use warnings;
use strict;
$|=1;
use Carp;
use Data::Dumper;

use File::Path      qw( mkpath );
use File::Basename  qw( fileparse basename dirname );
use File::Copy      qw( copy move );
use autodie         qw( :all mkpath copy move ); # system/exec along with open, close, etc
use Cwd             qw( cwd abs_path );
use Env             qw( HOME );
use List::MoreUtils qw( any );
use String::ShellQuote qw( shell_quote_best_effort );
use Config::Std;
use Getopt::Long    qw( :config no_ignore_case bundling );
use List::Util      qw( first max maxstr min minstr reduce shuffle sum );

use utf8::all;

our $VERSION = 0.01;
my  $prog    = basename($0);

my $DEBUG   = 0;
my $working_area = "/tmp/.skullplot";   # default
my ($dependent_spec, $independent_spec, $indie_count, $image_viewer);

GetOptions ("d|debug"    => \$DEBUG,
            "v|version"  => sub{ say_version(); },
            "h|?|help"   => sub{ say_usage();   },

           "dependents=s"     => \$dependent_spec,    # the x-axis, plus any gbcats
           "independents=s"   => \$independent_spec,  # the y-axis

           "indie_count=i" => \$indie_count,          # alt spec: indies=x1+gbcats; residue are ys

           "image_viewer=s" => \$image_viewer,        # default: ImageMagick's display (if available)

           "working_area=s"   => \$working_area,
           ) or say_usage();


unless( -d $working_area ) {
  mkpath( $working_area );
}


# TODO for development only.  Remove when shipped.
use FindBin qw( $Bin );
use lib ("$Bin/../lib/");

use Data::BoxFormat;

my $dbox_file = shift;

unless( $dbox_file ) {
  die "An input data file (*.dbox) is required.";
}

if( $dependent_spec && not( $independent_spec ) ) {
  die "When using dependents option, also need independents.";
} elsif( $independent_spec && not( $dependent_spec ) ) {
  die "When using independents option, also need dependents.";
} elsif( $indie_count && $dependent_spec) {
  die "Use either indie_count or dependents/independents options, not both.";
}

if ( $dependent_spec ) {
  ($DEBUG) &&
    print STDERR "Using independents: $independent_spec and dependents: $dependent_spec\n";
} elsif( $indie_count )  {
  ($DEBUG) &&
    print STDERR "Given indie_count: $indie_count\n";
} else {
  $indie_count = 1;
}

### generate a tsv file in the working area, along with rscript

my $dbox_name = basename( $dbox_file );
( my $tsv_name     = $dbox_name ) =~ s{ \.dbox $ }{.tsv}x;
( my $rscript_name = $dbox_name ) =~ s{ \.dbox $ }{.t}x;

my $tsv_file =     "$working_area/$tsv_name";
my $rscript_file = "$working_area/$rscript_name";

($DEBUG) && print "input dbox name: $dbox_name\nintermediate tsv_file: $tsv_file\n";

# input from dbox file, output directly to a tsv file
my $dbx = Data::BoxFormat->new( input_file  => $dbox_file,
                            output_file => $tsv_file,
                          );
$dbx->scrape2tsv();

# use first col as the default independent variable (x-axis)
my @header = @{ $dbx->header() };
my $independent_default  = shift( @header );

# my $dependent_default    = join ',', @header;

my ($dep_list, $indep_list);
my ( @dep_fields, @indep_fields, @gb_cats, $x_axis );
if( $indie_count ) {
  $x_axis = $independent_default;
  @gb_cats      = @header[ 0 .. $indie_count-1 ];
  @indep_fields = @header[ $indie_count .. $#header ];
} elsif ( $dependent_spec || $independent_spec )  {
  $dep_list     = $dependent_spec   || join ',', @header;
  $indep_list   = $independent_spec || $independent_default;
  @dep_fields   = split /[,\|]/, $dep_list;
  @indep_fields = split /[,\|]/, $indep_list;
  $x_axis       = shift( @indep_fields );
  @gb_cats      = @dep_fields;
}

my $y_field = $indep_fields[ 0 ];  ### TODO DEBUG just ploting the first to start with
print STDERR "x_axis: $x_axis\n";
print STDERR "y_field: $y_field\n";
# my $plot_code = 'qplot( skull$' . $x_axis . ', skull$' . $y_field . ' )';

# TODO
# At present, limited to two group by categories (in addition to the x-axis)
# Maybe: if there's more than 2, fuse them together into a compound cat, hand to colour
my $gb_cat1 = $gb_cats[0] if $gb_cats[0];
my $gb_cat2 = $gb_cats[1] if $gb_cats[1];

# plot code
my $pc = 'ggplot( skull, ' ;
  $pc .= '               aes(' ;
  $pc .= "                    x = $x_axis," ;
  $pc .= "                    y = $y_field, " ;
  $pc .= "                    colour = $gb_cat1," if $gb_cat1;
  $pc .= "                    shape  = $gb_cat2 " if $gb_cat2;
  $pc .= '                          ))' ;
  $pc .= ' + geom_point( ' ;
#  $pc .=  "             shape = 21, " ;
  $pc .= "              size  = 2.5 " ;
  $pc .= '              )  ' ;

(my $png_file = $tsv_file ) =~s{ \.tsv $ }{.png}x;

### Generate R code file to run with Rscript call
### (in debug mode, make it a standalone unix script)
my $r_code;
$r_code = qq{#!/usr/bin/Rscript} . "\n" if $DEBUG;

$r_code .=<<"END";
library(ggplot2)
skull <- read.delim("$tsv_file", header=TRUE)
png("$png_file") # send plot output to png
$pc
graphics.off()   # doesn't chatter like dev.off
END

open my $out_fh, '>', $rscript_file;
print { $out_fh } $r_code;
close( $out_fh );

# in case you want to run the rscript standalone
chmod 0755, $rscript_file;

# chdir( "$HOME/tmp" ) or die "$!";

my $erroff = '2>/dev/null';
$erroff = '' if $DEBUG;

my $cmd = "Rscript $rscript_file $erroff";

print STDERR "cmd:\n$cmd\n" if $DEBUG;
system( $cmd );

  # TODO make choice of display app settable...
  #      have an intelligent default: look at what's available.

# Running a desktop app is faster than R's browseURL:
# system("gthumb $png_file    $erroff");
# system("eog $png_file       $erroff");
# system("display $png_file   $erroff");  # ImageMagick
# exec(qq{ display  $png_file $erroff }); # ImageMagick
# exec(qq{ display -title 'skullplot'  $png_file $erroff }); # ImageMagick

# (Defaulting to ImageMagick's "display" is good, because I can
# use a dependency on Perlmagick to ensure that it's available)

my $vcmd;
if( $image_viewer ) {
  $vcmd = qq{ $image_viewer $png_file $erroff };
} else {
  $vcmd = qq{ display -title 'skullplot'  $png_file $erroff };
}

exec( $vcmd );


#######
### end main, into the subs

sub say_usage {
  my $usage=<<"USEME";
  $prog -[options] [arguments]

  Options:
     -d          debug messages on
     --debug     same
     -h          help (show usage)
     -v          show version
     --version   show version

TODO add additional options

USEME
  print "$usage\n";
  exit;
}

sub say_version {
  print "Running $prog version: $VERSION\n";
  exit 1;
}


__END__

=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 BUGS

None reported... yet.

=cut


lib/Data/BoxFormat.pm
package Data::BoxFormat;
use Moo;
use MooX::Types::MooseLike::Base qw(:all);

=head1 NAME

Data::BoxFormat - work with db SELECT results in tabular text format

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';
my $DEBUG = 1;              # TODO revise before shipping

=head1 SYNOPSIS

   use Data::BoxFormat;
   # Reading input from a "dbox" temp file
   my $dbx = Data::BoxFormat->new( input_file => '/tmp/select_result.dbox' );
   my $data = $dbx->read(); # array of arrays, header in first row


   # Input from a string
   my $dbx = Data::BoxFormat->new( input_data => $dboxes_string );
   my $data = $dbx->read();


   # input from dbox file, output directly to a tsv file
   my $dbx = Data::BoxFormat->new( input_file  => '/tmp/select_result.dbox',
                               output_file => '/tmp/select_result.tsv',
                             );
   $dbx->read2tsv();


   # TODO test this
   # Use just a unicode character as sepatator via a custom
   # parsing regexp (this allows '|' in string values).
   my $dbx =
      Data::BoxFormat->new( input_file => '/tmp/select_result.dbox',
                        delimiter_re =>
                          sub{ qr{
                                \s+
                                 \N{BOX DRAWINGS LIGHT VERTICAL}
                                    {1,1}
                                \s+
                            }xms }
                       );
   my $data = $dbx->read(); # array of arrays, header in first row




=head1 DESCRIPTION

Data::BoxFormat is a module to work with data in the tabular text
format(s) commonly used in database client shells (postgresql's
"psql", mysql's "mysql", etc), where a SELECT will typical display
data in a form such as this (mysql):

  +-----+------------+---------------+-------------+
  | id  | date       | type          | amount      |
  +-----+------------+---------------+-------------+
  |  11 | 2010-09-01 | factory       |   146035.00 |
  |  12 | 2010-10-01 | factory       |   208816.00 |
  |  15 | 2011-01-01 | factory       |   191239.00 |
  |  16 | 2010-09-01 | marketing     |   467087.00 |
  |  17 | 2010-10-01 | marketing     |   409430.00 |
  +-----+------------+---------------+-------------+

Or this (postgresql's "ascii" form):

   id |    date    |   type    | amount
  ----+------------+-----------+--------
    1 | 2010-09-01 | factory   | 146035
    2 | 2010-10-01 | factory   | 208816
    4 | 2011-01-01 | factory   | 191239
    6 | 2010-09-01 | marketing | 467087
    7 | 2010-10-01 | marketing | 409430

These formats are human-readable, but not suitable for other
purposes such as feeding to a graphics program, or inserting into
another database table.

This code presumes these text tables of "data boxes" are either
held in a string or saved to a file.

By default this code can work with three different
formats: mysql, psql and unicode psql.

=head2 implementation notes

To write general code to work with all three formats, we
need to do three things:

(1) skip lines like this

  +-----+------------+---------------+-------------+
  ----+------------+-----------+--------
  ────┼────────────┼───────────┼────────

(2) treat either of these two characters as data delimiters,
  the ascii vertical bar or the unicode "BOX DRAWINGS LIGHT VERTICAL":
    |│

007C
007C;VERTICAL LINE;Sm;0;ON;;;;;N;VERTICAL BAR;;;;
|


(3) strip leading and trailing whitespace on each value

And we can cover some possible variations with some object-level
settings...

=head3 unicode characters

the unicode psql format uses these three characters:

uniprops U+2502
U+2502 ‹│› \N{BOX DRAWINGS LIGHT VERTICAL}
    \pS \p{So}
    All Any Assigned InBoxDrawing Box_Drawing Common Zyyy So S Gr_Base
       Grapheme_Base Graph GrBase Other_Symbol Pat_Syn Pattern_Syntax PatSyn
       Print Symbol Unicode X_POSIX_Graph X_POSIX_Print

uniprops U+2500
U+2500 ‹─› \N{BOX DRAWINGS LIGHT HORIZONTAL}
    \pS \p{So}
    All Any Assigned InBoxDrawing Box_Drawing Common Zyyy So S Gr_Base
       Grapheme_Base Graph GrBase Other_Symbol Pat_Syn Pattern_Syntax PatSyn
       Print Symbol Unicode X_POSIX_Graph X_POSIX_Print

uniprops U+253c
U+253C ‹┼› \N{BOX DRAWINGS LIGHT VERTICAL AND HORIZONTAL}
    \pS \p{So}
    All Any Assigned InBoxDrawing Box_Drawing Common Zyyy So S Gr_Base
       Grapheme_Base Graph GrBase Other_Symbol Pat_Syn Pattern_Syntax PatSyn
       Print Symbol Unicode X_POSIX_Graph X_POSIX_Print




=head1 METHODS

=over

=cut

use 5.008;
use Carp;
use Data::Dumper;
use utf8::all;

=item new

Creates a new Data::BoxFormat object.

Takes a list of attribute/setting pairs as an argument.

=over

=item input_file

File to input data from.  Only required if L<input_data> was not
defined directly.

=item input_encoding

Default's to "UTF-8".  Change to suit text encoding (e.g. "ISO-8859-1").
Must work as a perl ":encoding(...)" layer.

=item input_data

SQL SELECT output in the fixed-width-plus-delimiter form discussed above.

=item output_file

Like L<input_file>, only used by L<read2tsv> (at present)

=item output_encoding

Like L<input_encoding>.  Default: "UTF-8".

=item the parsing regular expressions (type: RegexpRef)

=over

=item delimiter_re

The column separators (vertical bar)
TODO rename this: separator, not delimiter.

=item horizontal_re

Horizontal bars inserted for readability

=item cross_re

Match cross marks the horizontal bars typically use to mark
column boundaries (not yet in use).

=item left_edge_re

Left border delimiters (we strip these before processing).

=item right_edge_re

Right border delimiters (we strip these before processing).

=back

=back

=cut

# Example attribute:
# has is_loop => ( is => 'rw', isa => Int, default => 0 );

# input file name (can skip if input_data is defined directly)
has input_file  => ( is => 'rw', isa => Str, default => "" );

# input encoding defaults to utf-8 (might need to change, e.g. ISO-8859-1)
has input_encoding => ( is => 'rw', isa => Str, default => 'UTF-8' );

# only used by read2tsv (at present)
has output_file     => ( is => 'rw', isa => Str, default => "" );
### TODO lazy default, generate from the input_file, change extension to tsv

has output_encoding => ( is => 'rw', isa => Str, default => 'UTF-8' );

# can define input data directly, or alternately slurp it in from a file
has input_data  => ( is => 'rw', isa => Str,
           default =>
             sub { my $self = shift;
                   my $input_file = $self->input_file;
                   my $input_encoding = $self->input_encoding;
                    unless ( $input_file ) {
                     croak
                       "Needs either an input data file name ('input_file'), " .
                       "or a multiline string ('input_data')  ";
                   }
                    my $in_enc = "<:encoding($input_encoding)";
                   open my $fh, $in_enc, $input_file or croak "$!";
                   local $/; # localized slurp mode
                   my $data = <$fh>;
                   return $data;
                 },
            lazy => 1 );
# TODO would be better if it avoided slurping, and worked line-at-a-time

has header => ( is => 'rw', isa => ArrayRef, default => sub{ [] } );

# defining character properties for regexp defaults
sub IsHor {
  my @codepoints =
    ('002D',  # -  \N{HYPHEN-MINUS}
     '002B',  # +  \N{PLUS SIGN}
     '2500',  # ─  \N{BOX DRAWINGS LIGHT HORIZONTAL}
     '253C',  # ┼  \N{BOX DRAWINGS LIGHT VERTICAL AND HORIZONTAL}
     );
  my $list = join "\n", @codepoints;
  return $list;
}

sub IsCross {
  my @codepoints =
    (
     '002B',  # +  \N{PLUS SIGN}
     '253C',  # ┼  \N{BOX DRAWINGS LIGHT VERTICAL AND HORIZONTAL}
     );
  my $list = join "\n", @codepoints;
  return $list;
}

sub IsDelim {
  my @codepoints =
    (
     '007C',  # |  \N{VERTICAL LINE}
     '2502',  #  │  \N{BOX DRAWINGS LIGHT VERTICAL}
    );
  my $list = join "\n", @codepoints;
  return $list;
}

# TODO
# but don't *really* want to allow mix-and-match delims, it's always one or the other throughout.
has delimiter_re  => ( is => 'rw', isa => RegexpRef,
                       default =>
                       sub{ qr{
                                \s+         # require leading whitespace
                                \p{IsDelim}
                                    {1,1}   # just one delim char
                                \s+         # require trailing whitespace
                            }xms } );


has horizontal_re => ( is => 'rw', isa => RegexpRef,
                       default => sub{ qr{ \p{IsHor}+ }x } );

# TODO
#   horizontal rules have crosses that mark-the-spots of col borders
#   use this piece of information: fixed width fields, delims are lined-up
has cross_re  => ( is => 'rw', isa => RegexpRef,
                       default =>
                       sub{ qr{
                                \p{IsCross}
                                    {1,1}   # just one delim char
                            }xms } );

# To match table borders (e.g. mysql-style)
has left_edge_re => ( is => 'rw', isa => RegexpRef,
                       default => sub{ qr{ ^ \s* [\|] }xms } );

has right_edge_re => ( is => 'rw', isa => RegexpRef,
                       default => sub{ qr{ [\|] \s* $ }xms } );


=item read

Given data in tabular boxes from a multiline string,
convert it into an array of arrays.

   my $data =
         $bxs->read();

Goes through the boxdata slurped into the object field input_data,
returns it as an array of arrays, including the field names in
the first row.

As a side-effect, stores the header (first row of boxdata)
in the object's L<header>.

=cut


### TODO add features to look for bad data formats
### TODO even better: be more forgiving of values with embedded delim chars.
sub read {
  my $self = shift;

  my $input_data    = $self->input_data;
  my $horizontal_re = $self->horizontal_re;  # intended to skip these... TODO
  my $delimiter_re  = $self->delimiter_re;

  # before we split on delimiters, trim the left and right borders (if any)
  # (converts mysql-style lines into psql-style lines)
   my $left_edge_re  = $self->left_edge_re;
   my $right_edge_re = $self->right_edge_re;

   $input_data =~ s{ ^ \s* \+       }{}xmsg;
   $input_data =~ s{ \+ \s* $       }{}xmsg;
   $input_data =~ s{ $left_edge_re  }{}xmsg;
   $input_data =~ s{ $right_edge_re }{}xmsg;

  # Here we just look for lines with delimiters on them (skipping
  # anything else) and then split the lines on the delimiters,
  # trimming whitespace from the boundaries of all values

  my @lines = split /\n/, $input_data;

  my ( @data );
  for my $i ( 0 .. $#lines ) {
    my $line = $lines[ $i ];

    # when there's at least one delim, we assume it's a data line
    if( $line =~ /$delimiter_re/xms ) {

      no warnings 'uninitialized';

      # need this for the whitespace not adjacent to delimiters...
      $line =~ s/^\s+//; # strip leading spaces
      $line =~ s/\s+$//; # strip trailing spaces (if any)

      # Note: split pattern also eats bracketing whitespace
      my @vals =
        split /$delimiter_re/, $line;

      # array_of_array format (header treated like any other vals)
      push @data, \@vals;
    }

    my @header;
    @header = @{ $data[0] } if @data;
    # print STDERR Dumper( \@header ), "\n"; # TODO when turned on there's an odd instance in tests where an empty one comes up
    $self->header( \@header );
  }
  return \@data;
}



=item read2tsv

A convenience method that runs L<read> and writes the data to a
tsv file specified by the field L<output_file>.

=cut

sub read2tsv {
  my $self = shift;

  my $output_file = $self->output_file;
  unless( $output_file ) {
    croak("read2tsv requires the output_file.");
    ## TODO could use the input_file (if defined) changing the extension to tsv.
  }
  my $output_encoding = $self->output_encoding;

  my $data = $self->read;

  my $out_enc = ">:encoding($output_encoding)";
  open my $fh, $out_enc, $output_file or die "$!";

  for my $i ( 0 .. $#{ $data } ) {
    my $line = join "\t", @{ $data->[ $i ] };
    print { $fh }  $line, "\n";
  }
  return 1;
}




=back

=head1 AUTHOR

Joseph Brenner, E<lt>doom@kzsu.stanford.eduE<gt>,
05 Jun 2016

=head1 BUGS

Please report any bugs or feature requests to
C<bug-data-boxes at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-BoxFormat>.

I will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Emacs::Run

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Emacs-Run>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Emacs-Run>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Emacs-Run>

=item * Search CPAN

L<http://search.cpan.org/dist/Emacs-Run/>

=back

=cut

=head1 ACKNOWLEDGEMENTS

I stand on the shoulders of leprechauns.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Joseph Brenner

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
